core/base/inc/TDirectory.h

```diff
@@ -299,57 +324,54 @@ can be replaced with the simpler and exception safe:
    ClassDefOverride(TDirectory,5)  //Describe directory structure in memory
 };
 
-#ifndef __CINT__
 namespace ROOT {
 namespace Internal {
    struct TDirectoryAtomicAdapter {
-      std::atomic<TDirectory*> &fValue;
-      TDirectoryAtomicAdapter(std::atomic<TDirectory*> &value) : fValue(value) {}
+      TDirectory::SharedGDirectory_t &fValue;
```

Comment:
> it might make sense to include the main user-facing interface

I am not sure what you mean ...

> How can you know that the referenced shared_ptr has a sufficient lifetime?

The lifetime of the `shared_ptr` is that of the thread.  The `TDirectoryAtomicAdapter` are intended to be *"temporary-only*" object.  Indeed it would be a disaster if the user uses this *Internal* type as a `thread_local` or `static` variable or anything that might have a lifetime longer than the thread that created it.

Reply:
> The lifetime of the shared_ptr is that of the thread.

That's super useful information, thanks, I missed that! Maybe worth calling out in a comment?

> The TDirectoryAtomicAdapter are intended to be "temporary-only" object.

I see the following code (line 376):
```c++
#define gDirectory (ROOT::Internal::TDirectoryAtomicAdapter{})
```

So `TDirectoryAtomicAdapter` is user facing, because `gDirectory` is commonly used.

Is
```c++
auto myDirToBePassedToOtherThread = gDirectory;
```
 fine? I guess the key here is the passing-to-other-thread, which would always happen as `TDirectory*` and not `auto`?

Can this line be problematic if there is task nesting?

Reply:
And btw
```c++
#define gDirectory (ROOT::Internal::TDirectoryAtomicAdapter{})
```
also makes the `Internal` namespace a bit dubious...